accepted/2023/net8.0-browser-tfm.md

```diff
@@ -0,0 +1,187 @@
+# net8.0-browser TFM for applications running in the browser
+
+**Owner** [Javier Calvarro](https://github.com/javiercn) | [Javier Calvarro](https://github.com/javiercn)
+**Contact** [Daniel Roth](https://github.com/danroth27) | [Daniel Roth](https://github.com/danroth27)
+**Contact** [Artak Mkrtchyan](https://github.com/mkartakmsft) | [Artak Mkrtchyan](https://github.com/mkartakmsft)
+
+For a video introduction to what we are trying to achieve, see:
+https://www.youtube.com/live/kIH_Py8ytlY?feature=share&t=473
+https://www.youtube.com/watch?v=48G_CEGXZZM
+
+We have been shipping .NET in the browser via webassembly since the introduction of Blazor 3.2.0. During that initial release we decided to avoid adding a TFM for the browser as we were not clear on whether we needed it and we knew it was a one way operation. Once we introduced a TFM, we would never be able to take it back.
+
+Our reasoning was that we could likely get away with a RID as that would allow third-parties to provide alternative pre-compiled webassembly assets (including native dependencies compiled to wasm).
+
+We reasoned that we did not need a TFM as we could always provide a PAL over any functionality we needed and we would annotate the APIs with the `[SupportedOSPlatform]` attribute and an analyzer that would warn against incorrect usage when targeting the browser.
+
+Over time, we have learned that the lack of a TFM is limiting for us and our customers. The lack of a TFM introduces friction in two ways:
+* **Exposing additional customer-facing APIs when running in browser (and not on desktop)**: For example, Blazor includes many platform agnostic abstractions like IJSRuntime, IJSInProcessRuntime and different implementations to be able to reflect the different capabilities of the different platforms. This makes taking advantage of some of the webassembly unique capabilities more challenging as this kind of functionality needs to be carefully designed, and creates other problems like lack of discoverability (you need to know the existence of IJSInProcessRuntime and downcast to it).
```

Comment:
I don't understand how having a single TFM would help with disjoint APIs sets here?

Reply:
Could you clarify what disjoint API sets you mean? In case you mean "some things are server-only, and some things are browser-only, and neither is a subset of the other" then that's true.

Here's an example of how things are in practice today:

 * `net7.0`
   * `System.Net.Sockets.Socket.Listen()` - works if you use it on server, throws at runtime if you try to use it in browser
   * `Microsoft.JSInterop.IJSInProcessRuntime` - works if you use it in browser, throws at runtime if you try to use it in server (i.e., when you try to downcast `IJSRuntime` to it)

... whereas here's how things *could* have been with a browser TFM:

 * `net7.0`
   * `System.Net.Sockets.Socket.Listen()` - works if you use it on server, throws at runtime if you try to use it in browser
 * `net7.0-browser`
   * `Microsoft.JSInterop.IJSInProcessRuntime` - works if you use it in browser. You can't accidentally try to use it in server because you're not targeting this TFM.

So, adding a browser TFM is not a 100% solution because TFMs just don't model this kind of scenario exactly. Adding the TFM solves *some* API misuse issues but not all of them. It's the best we can do with a TFM, and is an improvement over where we are today.

But as a recap, solving API misuse issues is not the main driving reason to do this - it's all the application model stuff Javier mentioned in this doc.

I hope this is what you're talking about. Apologies if I've completely misinterpreted your comment/question.

Reply:
This comment helped me understand that the main goal is to **allow** multiple DLLs for the same project.
Perhaps with `#if` sprinkled in the code.

And that would assure that client-side-only references would not be pulled into to server side deployment.

It would not prevent server-side-only dependencies to be referenced from the client side. Because `net8-browser` is superset of `net8`.

Reply:
> It would not prevent server-side-only dependencies to be referenced from the client side. Because net8-browser is superset of net8.

It wouldn't *prevent* them but it would give app developers and library developers the power to control that. They can mark a `PackageReference` or `ProjectReference` on `$(TargetFramework)` being exactly equal to `net8.0` or `net8.0-browser`, and then it won't be present in the other case.

Reply:
We should decorate `Microsoft.JSInterop.IJSInProcessRuntime` with `[SupportedOSPlatform("browser")]` either way. I also think that moving such type to to TFM only assembly at this point would be noticeable breaking change.

Reply:
@SteveSandersonMS 

> Could you clarify what disjoint API sets you mean?

The paragraph in spec says a TFM would help in offering WASM-specific APIs. A TFM is an API set. Multiple TFMs, disjoint API sets.

Historically, the problem of .NET was fragmentation. For library authors it's much easier to call APIs conditionally ("light-up") the building for multiple TFMs. The downside is discovery which APIs work on which platforms and bloat (you carry code that doesn't run). The platform analyzer addresses the former. The latter is a judgement call on our side which is why I asked earlier how many APIs we think will be WASM specific.

IOW, my question was why you believe that having multiple TFMs is making the UX easier.

Reply:
> I think what you're talking about is that by having multiple TFMs we avoid cases where folks accidentally call browser-specific APIs from the server. I was hoping that the built-in analyzer would have solved that, but it seems it didn't. Is that because customers turned it off or they don't understand it?

Thanks for clarifying.

If you're making a class library that is meant to support both server-side and WebAssembly execution modes, I'm not sure the platform compatibility analyzer really helps. You have to declare that you support `browser` (because you do), and at that point you lose all guarantees about what happens when you're running on the server. The PCA really only helps detect simple cases (you try to use an API that you definitely-can't-ever use, not when you unconditionally use an API that you sometimes-can-and-sometimes-can't).

Multitargeting over two TFMs correctly represents the scenario where you sometimes-can and sometimes-can't and forces you to account for this difference.

Reply:
That makes sense, thanks for explaining. I consider this resolved then.